File: plugins/pr-walkthrough/skills/pr-walkthrough/scripts/scaffold_site.py
#!/usr/bin/env python3
"""Create or refresh a PR walkthrough site from the reusable template."""

from __future__ import annotations

import argparse
import shutil
import subprocess
import sys
from pathlib import Path

OBSOLETE_TEMPLATE_PATHS = (
    "design-qa.md",
    "src/components/ui/card.tsx",
    "src/components/walkthrough/diff-browser.tsx",
    "src/components/walkthrough/graph-canvas.tsx",
)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--content",
        type=Path,
        help="Canonical walkthrough MDX directory. If omitted, the template includes sample Markdown.",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=Path(".pr-walkthrough/site"),
        help="Site source directory (default: .pr-walkthrough/site).",
    )
    parser.add_argument(
        "--diff",
        type=Path,
        help="Unified Git patch used for the changed-file tree and source diff view.",
    )
    parser.add_argument(
        "--include-full-context",
        action="store_true",
        help=(
            "Embed exact old/new Git blobs for native hunk expansion. "
            "The generated artifact must remain localhost-only."
        ),
    )
    args = parser.parse_args()

    template = Path(__file__).resolve().parent.parent / "assets" / "site-template"
    if not template.is_dir():
        raise SystemExit(f"template not found: {template}")

    if args.content and not args.content.is_dir():
        raise SystemExit("--content must be a directory containing index.mdx")
    if args.content and not (args.content / "index.mdx").is_file():
        raise SystemExit("--content must contain index.mdx")
    if args.diff and not args.diff.is_file():
        raise SystemExit(f"--diff file not found: {args.diff}")

    output = args.output.resolve()
    output.mkdir(parents=True, exist_ok=True)
    for relative_path in OBSOLETE_TEMPLATE_PATHS:
        stale_path = output / relative_path
        if stale_path.is_dir():
            shutil.rmtree(stale_path)
        else:
            stale_path.unlink(missing_ok=True)
    shutil.copytree(
        template,
        output,
        dirs_exist_ok=True,
        ignore=shutil.ignore_patterns("node_modules", ".next", "out", "__pycache__", "*.tsbuildinfo"),
    )

    if args.content:
        target = output / "src" / "content" / "walkthrough"
        shutil.rmtree(target, ignore_errors=True)
        shutil.copytree(args.content, target)

    patch_target = output / "src" / "data" / "walkthrough.patch"
    if args.diff:
        shutil.copyfile(args.diff, patch_target)
    elif args.content:
        patch_target.write_text("", encoding="utf-8")

    full_context_marker = output / "src" / "data" / "full-context.enabled"
    if args.include_full_context:
        full_context_marker.write_text("localhost-only\n", encoding="utf-8")
    else:
        full_context_marker.unlink(missing_ok=True)
        shutil.rmtree(output / ".next", ignore_errors=True)
        shutil.rmtree(output / "out", ignore_errors=True)

    try:
        subprocess.run(
            [sys.executable, "scripts/compile_walkthrough.py"],
            cwd=output,
            check=True,
        )
    except subprocess.CalledProcessError as error:
        raise SystemExit("walkthrough MDX compilation failed") from error

    print(output)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/pr-walkthrough/skills/pr-walkthrough/scripts/scaffold-site.ts
// Create or refresh a PR walkthrough site from the reusable template.
//
// Run with Bun:  bun run <skill-directory>/scripts/scaffold-site.ts [options]

import { spawnSync } from "node:child_process";
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

const OBSOLETE_TEMPLATE_PATHS = [
  "design-qa.md",
  "src/components/ui/card.tsx",
  "src/components/walkthrough/diff-browser.tsx",
  "src/components/walkthrough/graph-canvas.tsx",
];

const IGNORE_PATTERNS = ["node_modules", ".next", "out", "*.tsbuildinfo"];

const USAGE = `usage: scaffold-site.ts [-h] [--content CONTENT] [--output OUTPUT] [--diff DIFF]
                        [--include-full-context]

Create or refresh a PR walkthrough site from the reusable template.
`;

function usageError(message: string): never {
  process.stderr.write(USAGE);
  process.stderr.write(`scaffold-site.ts: error: ${message}\n`);
  process.exit(2);
}

function fail(message: string): never {
  process.stderr.write(`${message}\n`);
  process.exit(1);
}

function parseCliArgs(
  argv: string[],
  optionNames: string[],
  flagNames: string[],
): { values: Map<string, string>; flags: Set<string> } {
  const values = new Map<string, string>();
  const flags = new Set<string>();
  let index = 0;
  while (index < argv.length) {
    const token = argv[index];
    if (token === "-h" || token === "--help") {
      process.stdout.write(USAGE);
      process.exit(0);
    }
    const equals = token.indexOf("=");
    const name = equals >= 0 ? token.slice(0, equals) : token;
    if (equals < 0 && flagNames.includes(name)) {
      flags.add(name);
      index += 1;
      continue;
    }
    if (optionNames.includes(name)) {
      if (equals >= 0) {
        values.set(name, token.slice(equals + 1));
        index += 1;
        continue;
      }
      const value: string | undefined = argv[index + 1];
      if (value === undefined) usageError(`argument ${name}: expected one argument`);
      values.set(name, value);
      index += 2;
      continue;
    }
    usageError(`unrecognized arguments: ${token}`);
  }
  return { values, flags };
}

function isFile(target: string): boolean {
  try {
    return fs.statSync(target).isFile();
  } catch {
    return false;
  }
}

function isDirectory(target: string): boolean {
  try {
    return fs.statSync(target).isDirectory();
  } catch {
    return false;
  }
}

/** Resolve like Python's Path.resolve(): absolute, with symlinks followed. */
function resolveRealPath(target: string): string {
  const absolute = path.resolve(target);
  try {
    return fs.realpathSync(absolute);
  } catch {
    const parent = path.dirname(absolute);
    if (parent === absolute) return absolute;
    return path.join(resolveRealPath(parent), path.basename(absolute));
  }
}

function globMatch(pattern: string, name: string): boolean {
  const source = `^${pattern
    .replace(/[.+^${}()|[\]\\]/g, "\\$&")
    .replace(/\*/g, ".*")
    .replace(/\?/g, ".")}$`;
  return new RegExp(source).test(name);
}

function isIgnored(name: string): boolean {
  return IGNORE_PATTERNS.some((pattern) => globMatch(pattern, name));
}

function copyFileWithMetadata(source: string, destination: string): void {
  fs.copyFileSync(source, destination);
  const stats = fs.statSync(source);
  fs.chmodSync(destination, stats.mode);
  fs.utimesSync(destination, stats.atime, stats.mtime);
}

/** Recursive copy that merges into an existing tree, like shutil.copytree. */
function copyTree(source: string, destination: string, filterNames: boolean): void {
  fs.mkdirSync(destination, { recursive: true });
  for (const entry of fs.readdirSync(source, { withFileTypes: true })) {
    if (filterNames && isIgnored(entry.name)) continue;
    const from = path.join(source, entry.name);
    const to = path.join(destination, entry.name);
    if (entry.isDirectory()) copyTree(from, to, filterNames);
    else copyFileWithMetadata(from, to);
  }
  const stats = fs.statSync(source);
  fs.utimesSync(destination, stats.atime, stats.mtime);
}

function main(argv: string[]): number {
  const parsed = parseCliArgs(
    argv,
    ["--content", "--output", "--diff"],
    ["--include-full-context"],
  );
  const contentArg = parsed.values.get("--content");
  const outputArg = parsed.values.get("--output") ?? ".pr-walkthrough/site";
  const diffArg = parsed.values.get("--diff");
  const includeFullContext = parsed.flags.has("--include-full-context");

  const skillDir = path.dirname(path.dirname(resolveRealPath(fileURLToPath(import.meta.url))));
  const template = path.join(skillDir, "assets", "site-template");
  if (!isDirectory(template)) fail(`template not found: ${template}`);

  if (contentArg !== undefined && !isDirectory(contentArg)) {
    fail("--content must be a directory containing index.mdx");
  }
  if (contentArg !== undefined && !isFile(path.join(contentArg, "index.mdx"))) {
    fail("--content must contain index.mdx");
  }
  if (diffArg !== undefined && !isFile(diffArg)) fail(`--diff file not found: ${diffArg}`);

  const output = resolveRealPath(outputArg);
  fs.mkdirSync(output, { recursive: true });
  for (const relativePath of OBSOLETE_TEMPLATE_PATHS) {
    fs.rmSync(path.join(output, relativePath), { recursive: true, force: true });
  }
  copyTree(template, output, true);

  if (contentArg !== undefined) {
    const target = path.join(output, "src", "content", "walkthrough");
    fs.rmSync(target, { recursive: true, force: true });
    copyTree(contentArg, target, false);
  }

  const patchTarget = path.join(output, "src", "data", "walkthrough.patch");
  if (diffArg !== undefined) fs.copyFileSync(diffArg, patchTarget);
  else if (contentArg !== undefined) fs.writeFileSync(patchTarget, "", "utf8");

  const fullContextMarker = path.join(output, "src", "data", "full-context.enabled");
  if (includeFullContext) {
    fs.writeFileSync(fullContextMarker, "localhost-only\n", "utf8");
  } else {
    fs.rmSync(fullContextMarker, { force: true });
    fs.rmSync(path.join(output, ".next"), { recursive: true, force: true });
    fs.rmSync(path.join(output, "out"), { recursive: true, force: true });
  }

  const compile = spawnSync(process.execPath, ["run", "scripts/compile-walkthrough.ts"], {
    cwd: output,
    stdio: "inherit",
  });
  if (compile.status !== 0) fail("walkthrough MDX compilation failed");

  process.stdout.write(`${output}\n`);
  return 0;
}

process.exit(main(process.argv.slice(2)));
